include/math/dl/nn/engine.h

```diff
@@ -0,0 +1,444 @@
+    //DEBUG_PRINT("RETURN MULTI\n");
+    return out;
+  }
+
+  std::shared_ptr<Value<T>> pow(T other) {
+    // assert types
+    auto val = (T)std::pow(data, other);
+    std::string name = ("**" + std::to_string(other));
+    auto out = std::make_shared<Value<T>>(val, OP::OP_POW, "**", this->shared_from_this());
+
+    if(!is_param){
+      std::function<void()> backward = [out,other](){
+        DEBUG_PRINT_1("%spow %f -> ", BLUE, out->left->grad);
+        out->left->grad += other * std::pow(out->left->data, (other - static_cast<T>(1.0))) * out->grad;
+        DEBUG_PRINT_1("%f%s\n", out->left->grad, RESET);
+      };
+      out->backward_ = backward;
+    }
+    return out;
+  }
+
+  std::shared_ptr<Value<T>> exp(){
+    auto x = data;
+    std::vector<std::shared_ptr<Value<T>>> children = {this->shared_from_this()};
+    auto val = (T)std::exp(x);
+    auto out = std::make_shared<Value<T>>(val, OP::OP_EXP, "exp", this->shared_from_this());
+
+    if(!is_param){
+      std::function<void()> backward = [out](){
+        DEBUG_PRINT_1("%sexp %f -> ", BLUE, out->left->grad);
+        out->left->grad  += out->data * out->grad;
+        DEBUG_PRINT_1("%f%s\n", out->left->grad, RESET);
+      };
+      out->backward_ = backward;
+    }
+    return out;
+  }
+
+  std::shared_ptr<Value<T>> tanh(){
+    auto x = data;
+    std::vector<std::shared_ptr<Value<T>>> children = {this->shared_from_this()};
+    auto val = std::tanh(x);
+    auto out = std::make_shared<Value<T>>(val, OP::OP_TANH, "tanh", this->shared_from_this());
+
+    if(!is_param){
+      std::function<void()> backward = [out](){
+        DEBUG_PRINT_1("%stanh %f -> ", BLUE, out->left->grad);
+        out->left->grad  += static_cast<T>(1.0 - std::pow(out->data, 2.)) * out->grad;
+        DEBUG_PRINT_1("%f%s\n", out->left->grad, RESET);
+      };
+      out->backward_ = backward;
+    }
+    return out;
+  }
+
+  std::shared_ptr<Value<T>> relu(){
+    std::vector<std::shared_ptr<Value<T>>> children = {this->shared_from_this()};
+    T val = ((this->data) < static_cast<T>(0)) ? static_cast<T>(0.0) : this->data;
+    auto out = std::make_shared<Value<T>>(val, OP::OP_RELU, "ReLU", this->shared_from_this());
+    if(!is_param){
+      std::function<void()> backward = [out](){
+        DEBUG_PRINT_1("%sReLU %f -> ", BLUE, out->left->grad);
+        out->left->grad += static_cast<T>(out->left->data > static_cast<T>(0)) * out->grad;
+        DEBUG_PRINT_1("%f%s\n",out->left->grad, RESET);
+      };
+      out->backward_ = backward;
+    }
+    return out;
+  }
+
+  void backward() {
+    std::vector<std::shared_ptr<Value<T>>> topo;
+    std::set<std::shared_ptr<Value<T>>> visited;
+    std::function<void(std::shared_ptr<Value<T>>)> build_topo = [&](std::shared_ptr<Value<T>> v) {
+      if (visited.find(v) == visited.end()) {
+        visited.insert(v);
+        if(v->left != nullptr) {
+          build_topo(v->left);
+        }
+        if(v->right != nullptr) {
+          build_topo(v->left);
```

Comment:
```suggestion
          build_topo(v->right);
```
Oh my god... This is embarrassing... 3 days of debugging 🤦‍♂️

Reply:
<div>
  <iframe width="560" height="315" 
            src="https://www.youtube.com/embed/h4UqMyldS7Q" 
            title="YouTube video player" 
            frameborder="0" 
            allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" 
            allowfullscreen>
    </iframe>
</div>